src/core/coap/coap_message.cpp

```diff
@@ -177,6 +177,37 @@ Error Message::AppendOption(uint16_t aNumber, uint16_t aLength, const void *aVal
     return error;
 }
 
+Error Message::AppendOptionFromMessage(uint16_t aNumber, uint16_t aLength, const Message &aMessage, uint16_t aOffset)
```

Comment:
This new method contains a lot of repeated code as in `Error Message::AppendOption(uint16_t aNumber, uint16_t aLength, const void *aValue)`.

May be better to reactor the common code in a `private` method that can then be used by this and the exiting methods.

Reply:
Assume `Message::AppendOption` executes A,B,C and D，`Message::AppendOptionFromMessage` is different in C. I can only reactor the B part. I'd like to know better idea.

Reply:
Looks like this block is also common:
```
    VerifyOrExit(aNumber >= GetHelpData().mOptionLast, error = kErrorInvalidArgs);
    delta = aNumber - GetHelpData().mOptionLast;
```
Can we move this in  `AppendOptionHeader()`?

Reply:
I'm sorry I can't. This block is using `GetHelpData().mOptionLast` to check `aNumber` and calculate `delta`. And in the end of `Message::AppendOption`, `GetHelpData().mOptionLast` is updated. These two parts are corresponding. I can't separate the former and hide it in another help function.
In my opinion, the common code cannot always be wrapped into a function, either because a complete function is not being implemented, or because the scope of the complete function is exceeded ( only part of it can be wrapped ). On the other hand, help function is always a solution. 
In this case, it's better that this block remains in  `Message::AppendOption`.

Reply:
The line `GetHelpData().mOptionLast = aNumber;` can also be moved to new common function.

If an error happens during `Append`, the message (and all its state) should be considered invalid and malformed and should be discarded, so no reason to wait till end of the method to update `GetHelpData().mOptionLast`.